File: scripts/defect_categories.py
"""
Defect category configuration for reliability filtering.

This module defines which defects are considered "wear-and-tear" (consumables that
naturally wear out regardless of car quality) vs "reliability" (indicating actual
mechanical or structural issues with the vehicle).

Categorization is based on the `gebrek_artikel_nummer` field from RDW's gebreken dataset.
"""

from typing import Any

# Article number patterns that indicate WEAR-AND-TEAR defects (NOT reliability issues)
# These are consumables/items that naturally wear out regardless of car quality:
WEAR_AND_TEAR_ARTICLE_PATTERNS: set[str] = {
    # Tires / Banden (profiel, montage, etc.) - Art 5.*.27
    "5.*.27",
    "5.2.27",
    "5.3.27",
    # Tire-related advisory codes
    "TA.5C",
    "TA.5D",
    "TA.6G",
    "TA.6H",
    "AC1",  # Advisory: tire profile 1.6-2.5mm
    # Lights / Verlichting - Bulbs wear out over time
    "5.*.51",
    "5.*.52",
    "5.*.53",
    "5.*.55",
    "5.*.61",
    "5.*.62",
    "5.2.51",
    "5.2.52",
    "5.2.53",
    "5.2.55",
    "5.3.51",
    "5.3.52",
    "5.3.53",
    "5.3.55",
    # Windshield wipers - normal wear items
    "5.*.36",
    "5.2.36",
    "5.3.36",
}

# Defect code prefixes that indicate wear-and-tear (administrative/advisory codes)
WEAR_AND_TEAR_CODE_PREFIXES: tuple[str, ...] = (
    "AC",  # Advisory Codes
    "AP",  # APK+ codes (advisory)
)

# Specific defect codes that are NOT reliability indicators
# (These override article number matching for edge cases)
WEAR_AND_TEAR_SPECIFIC_CODES: set[str] = {
    "205",  # Band onvoldoende profiel
    "206",  # Band onvoldoende profiel (<1.4mm)
    "212",  # Tire pressure warning
    "213",  # Tire mounting not per spec
    "216",  # Load index too small
    "217",  # Load index not determinable
    "701",  # Tire profile advisory
    "875",  # Bandenprofiel just below minimum
    "876",  # Bandenprofiel insufficient
    "RA4",  # Tire pressure system warning
    "AC1",  # Tire profile advisory
}


def is_wear_and_tear_defect(
    defect_code: str,
    gebreken_index: dict[str, dict[str, Any]] | None = None,
) -> bool:
    """
    Check if a defect code represents wear-and-tear rather than reliability.

    Args:
        defect_code: The gebrek_identificatie code
        gebreken_index: Optional index of defect descriptions (for article number lookup)

    Returns:
        True if this defect is wear-and-tear, False if it's a reliability indicator
    """
    defect_code = defect_code.strip().upper()

    # Check if it's in the specific wear-and-tear codes
    if defect_code in WEAR_AND_TEAR_SPECIFIC_CODES:
        return True

    # Check prefix-based exclusions
    for prefix in WEAR_AND_TEAR_CODE_PREFIXES:
        if defect_code.startswith(prefix):
            return True

    # If we have the gebreken index, check article number
    if gebreken_index:
        defect_info = gebreken_index.get(defect_code, {})
        article_number = defect_info.get("gebrek_artikel_nummer", "").strip()

        # Check if article number matches any wear-and-tear pattern
        if article_number in WEAR_AND_TEAR_ARTICLE_PATTERNS:
            return True

        # Also check with wildcards replaced
        # Convert "5.2.27" to "5.*.27" pattern for matching
        if article_number and len(article_number) >= 5:
            parts = article_number.split(".")
            if len(parts) >= 3:
                wildcard_pattern = f"{parts[0]}.*.{parts[2]}"
                if wildcard_pattern in WEAR_AND_TEAR_ARTICLE_PATTERNS:
                    return True

    return False


def is_reliability_defect(
    defect_code: str,
    gebreken_index: dict[str, dict[str, Any]] | None = None,
) -> bool:
    """
    Check if a defect code represents a true reliability issue.

    This is the inverse of is_wear_and_tear_defect().

    Args:
        defect_code: The gebrek_identificatie code
        gebreken_index: Optional index of defect descriptions (for article number lookup)

    Returns:
        True if this defect indicates a reliability issue
    """
    return not is_wear_and_tear_defect(defect_code, gebreken_index)


# Category names for UI display
CATEGORY_RELIABILITY = "reliability"
CATEGORY_WEAR_AND_TEAR = "wear_and_tear"


def categorize_defect(
    defect_code: str,
    gebreken_index: dict[str, dict[str, Any]] | None = None,
) -> str:
    """
    Return the category name for a defect code.

    Args:
        defect_code: The gebrek_identificatie code
        gebreken_index: Optional index of defect descriptions

    Returns:
        Either CATEGORY_RELIABILITY or CATEGORY_WEAR_AND_TEAR
    """
    if is_wear_and_tear_defect(defect_code, gebreken_index):
        return CATEGORY_WEAR_AND_TEAR
    return CATEGORY_RELIABILITY

File: web/app/lib/defect_categories.ts
/**
 * Client-side defect categorization for reliability filtering.
 *
 * Classification based on Consumer Reports, JD Power, ADAC, and ANWB
 * methodology: wear-and-tear items (tires, bulbs, wipers, brake pads,
 * shocks, exhaust corrosion) are excluded from reliability scoring.
 *
 * IMPORTANT: AC3 (roestschade / rust damage) is a RELIABILITY indicator
 * and is intentionally NOT in this list. We use individual AC codes
 * instead of an AC* prefix to preserve AC3.
 */

// -- Category: Tires / Banden --
// Normal wear consumables; replaced every 30k-60k km
const TIRE_CODES: readonly string[] = [
    "203", // Band beschadigd
    "204", // Band ernstig beschadigd
    "205", // Band onvoldoende profiel
    "206", // Band onvoldoende profiel (<1.4mm)
    "207", // Uitstulping band
    "208", // Bandenspanning wijkt >0.5 bar af
    "209", // Bandenspanning 1 as wijkt >0.5 bar af
    "210", // Bandenspanning niet op juiste waarde
    "211", // Bandenspanning 1 as niet gelijk
    "212", // Bandenspanning waarschuwing
    "213", // Bandmontage niet conform specificatie
    "214", // Nageprofileerd
    "215", // Banden 1 as niet dezelfde maat
    "216", // Laadindex te klein
    "217", // Laadindex niet vast te stellen
    "218", // Noodwiel gemonteerd
    "701", // Bandprofiel advies
    "874", // Bandenspanning wijkt 2 bar af
    "875", // Bandprofiel net onder minimum
    "876", // Bandprofiel onvoldoende
    "877", // Bandenmaat aangedreven as wijkt af
    "992", // Band bevat uitstekend metalen element
    "AC1", // Band profieldiepte 1.6-2.5mm (advisory)
    "RA4", // Bandenspanning controlesysteem waarschuwing
];

// -- Category: Lighting / Bulbs --
// Bulb failures and alignment are consumable maintenance items
// per Consumer Reports methodology. NOT wiring or electrical faults.
const LIGHTING_CODES: readonly string[] = [
    // Bulb failures ("werkt niet")
    "475", // Achterlicht werkt niet
    "487", // Achteruitrijlicht werkt niet
    "497", // Kentekenplaatverlichting werkt niet
    "507", // Dim-/grootlicht werkt niet
    "532", // Mistachterlicht werkt niet
    "556", // Remlicht werkt niet
    "571", // Richtingaanwijzer werkt niet
    "572", // Zijrichtingaanwijzer werkt niet
    "573", // Waarschuwingsknipperlichten werken niet
    "584", // Stadslicht werkt niet
    "589", // Zij-markeringslicht werkt niet
    "601", // Mistlicht voor werkt niet
    // Light alignment / adjustment
    "516", // Dimlicht onjuist afgesteld
    "517", // Dimlicht onjuist afgesteld (variant)
    "518", // Dimlicht verblindend boven nullijn
    "519", // Dimlicht verblindend
    "520", // Koplamp afwijkend lichtbeeld
    "521", // Koplamp afwijkend lichtbeeld (variant)
    "522", // Licht-donkerscheiding onduidelijk
    "524", // Lichtbeeld onjuist
    "598", // Mistlicht voor onjuist afgesteld
    // Light appearance
    "542", // Retroreflector gebrek
    "576", // Richtingaanwijzers kleur wijkt af
    "710", // Lichtbeeld onjuist (advies)
];

// -- Category: Wipers / Washers --
// Wiper blades and washer fluid are consumable per Consumer Reports, ADAC
const WIPER_CODES: readonly string[] = [
    "388", // Ruitenwisser werkt niet
    "389", // Ruitenwisser werkt niet goed
    "391", // Ruitenwisser onvoldoende zicht
    "392", // Ruitensproeier werkt niet
    "393", // Ruitensproeier werkt niet goed
];

// -- Category: Brake Wear --
// Pad and disc wear only. Brake SYSTEM failures (leaks 296/297,
// ABS 321-325, master cylinder 278/279, lines 303-308, 916)
// are RELIABILITY indicators and are NOT included here.
const BRAKE_WEAR_CODES: readonly string[] = [
    "289", // Schijf-/trommelrem loopt niet vrij
    "290", // Schijf-/trommelrem loopt onvoldoende vrij
    "311", // Remvoering raakt trommel/schijf
    "312", // Remschijf ernstig gecorrodeerd
    "313", // Remschijf slijtage breukrisico
    "AC5", // Mechanische delen remsysteem slijtage (advisory)
    "705", // Mechanische delen remsysteem slijtage (advisory, same as AC5)
];

// -- Category: Shock Absorbers --
// Wear item with 80k-150k km expected lifespan
const SHOCK_ABSORBER_CODES: readonly string[] = [
    "241", // Schokdemper werkt niet goed
    "AC2", // Schokdemper lekkage (advisory)
    "702", // Schokdemper lekkage (advisory, same as AC2)
];

// -- Category: Exhaust Wear --
// Corrosion and mounting wear; expected 5-8 year lifespan.
// Emission SYSTEM failures (DPF 140/141, codes 122-143) are
// RELIABILITY indicators and are NOT included here.
const EXHAUST_WEAR_CODES: readonly string[] = [
    "115", // Uitlaatsysteem niet gasdicht
    "116", // Uitlaatsysteem niet geheel gasdicht
    "117", // Uitlaatsysteem ondeugdelijk bevestigd
];

// -- Category: Administrative / Non-vehicle --
// Not actual vehicle defects; inspection limitations or minor advisories
const ADMINISTRATIVE_CODES: readonly string[] = [
    "AC4", // Stuur/fuseekogel slijtage <=1.0mm (minor wear advisory)
    "704", // Stuur/fuseekogel slijtage (advisory, same as AC4)
];

// Prefixes that indicate non-vehicle / administrative codes.
// AP: administrative prefix (no codes in actual data, harmless safeguard)
// OP: inspection limitation codes (not actual vehicle defects)
const WEAR_AND_TEAR_PREFIXES: readonly string[] = ["AP", "OP"];

/** Category names for documentation and UI */
export type WearAndTearCategory =
    | "tires"
    | "lighting"
    | "wipers"
    | "brake_wear"
    | "shock_absorbers"
    | "exhaust_wear"
    | "administrative";

interface CategoryInfo {
    readonly category: WearAndTearCategory;
    readonly label: string;
    readonly codes: readonly string[];
}

const CATEGORIES: readonly CategoryInfo[] = [
    { category: "tires", label: "Tires / Banden", codes: TIRE_CODES },
    { category: "lighting", label: "Lighting / Bulbs", codes: LIGHTING_CODES },
    { category: "wipers", label: "Wipers / Washers", codes: WIPER_CODES },
    { category: "brake_wear", label: "Brake Wear", codes: BRAKE_WEAR_CODES },
    { category: "shock_absorbers", label: "Shock Absorbers", codes: SHOCK_ABSORBER_CODES },
    { category: "exhaust_wear", label: "Exhaust Wear", codes: EXHAUST_WEAR_CODES },
    { category: "administrative", label: "Administrative", codes: ADMINISTRATIVE_CODES },
];

// Single flat Set for O(1) lookup performance
const WEAR_AND_TEAR_CODES: Set<string> = new Set(
    CATEGORIES.flatMap((c) => c.codes),
);

/**
 * Check if a defect code represents wear-and-tear rather than reliability.
 */
export function isWearAndTearDefect(defectCode: string): boolean {
    const code = defectCode.trim().toUpperCase();

    if (WEAR_AND_TEAR_CODES.has(code)) {
        return true;
    }

    for (const prefix of WEAR_AND_TEAR_PREFIXES) {
        if (code.startsWith(prefix)) {
            return true;
        }
    }

    return false;
}

/**
 * Check if a defect code represents a true reliability issue.
 */
export function isReliabilityDefect(defectCode: string): boolean {
    return !isWearAndTearDefect(defectCode);
}

/**
 * Categorize a defect code as reliability or wear-and-tear.
 */
export function categorizeDefect(defectCode: string): "reliability" | "wear_and_tear" {
    return isWearAndTearDefect(defectCode) ? "wear_and_tear" : "reliability";
}

/**
 * Get all wear-and-tear codes (sorted) for display or configuration.
 */
export function getDefaultWearAndTearCodes(): string[] {
    return Array.from(WEAR_AND_TEAR_CODES).sort();
}

/**
 * Get the wear-and-tear category breakdown for documentation or UI display.
 * Returns each category with its label and associated defect codes.
 */
export function getWearAndTearCategories(): readonly CategoryInfo[] {
    return CATEGORIES;
}
